### ISSUE TITLE
Message-based controls don't work in the console interface

### ISSUE THREAD

**Issue by jofemodo** [drobilla/jalv#62]
Hi @drobilla !

Regarding "property parameters" support in Jalv. What is the current status? Is supported in anyway?
I see there is some code for it but it doesn't seem to work in the plugins i've tested with.

**Background:**
As you probably know, Juce now export LV2 plugins using these extension, but we can't find too much info about it and LV2 specs are really "minimalistic" about it. Please, could you give me some tips about this or point me to more detailed documentation i can read? 

Thanks a lot!

**Comment by drobilla** [comment]
As far as I can tell it works with the eg-params example, as of about a decade ago.

I can't really say anything useful about "it doesn't seem to work".  Why not?  Compared to what?  What _does_ work?  What messages are being sent?  Are they being received?  What is different about this from the above-mentioned known working case? etc

**Comment by jofemodo** [comment]
Example. Using Surge XT. I load the plugin with jalv (cli) and then load some preset. The preset is loaded and you can play. When i print the controils from the command line, all of them are printed to 0. Trying to modify any parameter don't cause any effect and the parameter keeps  the 0 value.

```
root@zynthian:/usr/local/lib/lv2/Surge XT.lv2# jalv https://surge-synthesizer.github.io/lv2/surge-xt
Plugin:       https://surge-synthesizer.github.io/lv2/surge-xt
JACK Name:    Surge XT
Sample rate:  48000 Hz
Block length: 128 frames
MIDI buffers: 32768 bytes
Comm buffers: 1023872 bytes
Update rate:  30.0 Hz
Scale factor: 1.0

> preset https://surge-synthesizer.github.io/lv2/surge-xt:preset3004
surge_xt_b_lfo0_trigmode = 0.000000
surge_xt_b_lfo0_unipolar = 0.000000
surge_xt_b_lfo10_attack = 0.000000
surge_xt_b_lfo10_decay = 0.000000
surge_xt_b_lfo10_deform = 0.000000
surge_xt_b_lfo10_delay = 0.000000
surge_xt_b_lfo10_hold = 0.000000
surge_xt_b_lfo10_magnitude = 0.000000
surge_xt_b_lfo10_phase = 0.000000
surge_xt_b_lfo10_rate = 0.000000
surge_xt_b_lfo10_release = 0.000000
surge_xt_b_lfo10_shape = 0.000000
surge_xt_b_lfo10_sustain = 0.000000
surge_xt_b_lfo10_trigmode = 0.000000
surge_xt_b_lfo10_unipolar = 0.000000
surge_xt_b_lfo11_attack = 0.000000
surge_xt_b_lfo11_decay = 0.000000
surge_xt_b_lfo11_deform = 0.000000
surge_xt_b_lfo11_delay = 0.000000
surge_xt_b_lfo11_hold = 0.000000
surge_xt_b_lfo11_magnitude = 0.000000
surge_xt_b_lfo11_phase = 0.000000
surge_xt_b_lfo11_rate = 0.000000
surge_xt_b_lfo11_release = 0.000000
surge_xt_b_lfo11_shape = 0.000000
surge_xt_b_lfo11_sustain = 0.000000
surge_xt_b_lfo11_trigmode = 0.000000
surge_xt_b_lfo11_unipolar = 0.000000
surge_xt_b_lfo1_attack = 0.000000
surge_xt_b_lfo1_decay = 0.000000
surge_xt_b_lfo1_deform = 0.000000
surge_xt_b_lfo1_delay = 0.000000
surge_xt_b_lfo1_hold = 0.000000
surge_xt_b_lfo1_magnitude = 0.000000
surge_xt_b_lfo1_phase = 0.000000
surge_xt_b_lfo1_rate = 0.000000
surge_xt_b_lfo1_release = 0.000000
surge_xt_b_lfo1_shape = 0.000000
surge_xt_b_lfo1_sustain = 0.000000
surge_xt_b_lfo1_trigmode = 0.000000
surge_xt_b_lfo1_unipolar = 0.000000
surge_xt_b_lfo2_attack = 0.000000
surge_xt_b_lfo2_decay = 0.000000
surge_xt_b_lfo2_deform = 0.000000
surge_xt_b_lfo2_delay = 0.000000
surge_xt_b_lfo2_hold = 0.000000
surge_xt_b_lfo2_magnitude = 0.000000
surge_xt_b_lfo2_phase = 0.000000
surge_xt_b_lfo2_rate = 0.000000
surge_xt_b_lfo2_release = 0.000000
surge_xt_b_lfo2_shape = 0.000000
surge_xt_b_lfo2_sustain = 0.000000
....

```

**Comment by jofemodo** [comment]
I've been reading and testing some changes in the source code. For instance, i modified jalv_console code to call the jalv_set_control() function passing the controlID, instead of working with control ports only:

```
  } else if (sscanf(cmd, "set %u %f", &index, &value) == 2) {
    if (index < jalv->controls.n_controls) {
      ControlID* control = jalv->controls.controls[index];
      jalv_set_control(jalv, control, sizeof(float), jalv->urids.atom_Float, &value);
      jalv_print_control(jalv, control, value);
    } else {
      fprintf(stderr, "error: port index out of range\n");
    }
  }
```
This code compiles OK, but the "jalv_set_control()" function seems to do nothing with properties. No error, no action. Parameters remain unchanged. Curiously, the modified code still works with control ports, but not with properties.

Regards,

**Comment by drobilla** [comment]
I see.  There's support in the generic graphical UI, I wasn't thinking about the (openly toyish) command line interface.  As you've noticed, they don't seem to work there.

`jalv_set_control` is what's used by the graphical UIs and should work.

**Comment by jofemodo** [comment]
Let me do extra checks with `jalv_set_control`.

Regarding the reading of this kind of parameters, i understand that changes are "pushed" from the plugin and i've to "capture" the event. Indeed, i suppose it's already done for the GUI.
Do you think we could implement the zynthian interface with Jalv like a generic GUI module? It would be possible to extend jalv to manage 2 GUIs at once for the same plugin? If not, we are thinking of adding a OSC layer to jalv.

Regards,

**Comment by jofemodo** [comment]
Hi @drobilla !

It seems jalv_set_control()  works only when the plugin is run with a GUI. I checked with jalv.gtk3, generic and native GUI, and both work.
In the other hand, if the plugin is run with "jalv" executable (jalv_console.c), calling "jalv_set_control()" does nothing. I'm trying to understand and fix this in our fork.

In the other hand, when running the generic UI, control widgets are not updated with default values at startup nor after loading a preset. Initially they are all to zero, but when moved, the control works and the parameter value is changed. If a "jalv_set_control()" is sent (we have a jalv.gtk3 patched version with full console running in a thread), this also works, but again, widgets doesn't get updated.  Do you want i open a separated ticket for this?

Thanks

**Comment by jofemodo** [comment]
OK! I get it to work in the console gui-less. The "jalv_set_control()" function is using the "ui_to_plugin" ring-buffer and the "real" atoms are really sent by "jalv_apply_ui_events()", that was "disabled" if !jalv->has_ui.

So it's working half-way. Now i need to understand how updates work from plugin to UI.

**Comment by seanwayland** [comment]
Hi Jofemodo ! Can I help ? My life is back to normal after 6 months of changing job fun and games .

**Comment by drobilla** [comment]
> Do you think we could implement the zynthian interface with Jalv like a generic GUI module? It would be possible to extend jalv to manage 2 GUIs at once for the same plugin? If not, we are thinking of adding a OSC layer to jalv.

Jalv isn't really designed to be either multi-plugin or multi-UI, and it suffers from starting life as a small and simple example host.  That said, I think the best way to approach this kind of thing is to add a cleaner interface (conceptually more or a protocol) to UIs so that other ones can be implemented cleanly.  There's not much that can actually happen between the "core" and UI anyway, it's mostly just two rings of messages that can go either way.

OSC is one of many frontends that could be easily added then.  A less toyish console interface as well.  They're more or less the same thing with different "syntax" anyway.

I haven't thought about multiple GUIs specifically, or think this would be terribly useful/appropriate to be honest, but maybe it would be easy enough.  The most obvious practical goal of this would be to have the command interface and a GUI at the same time.  I don't think any of this would be terribly difficult, it just takes some doing.  I've been very slowly working towards cleaning things up vaguely in this direction, but actually doing all of this isn't very high on my priority list.

The key thing to aspire to here (and reason I don't like some of the messy ways Jalv has been embedded so far) is for that interface to be very small, and opaque.  It barely needs to do more than change some controls, and it should be correspondingly simple.

**Comment by drobilla** [comment]
> OK! I get it to work in the console gui-less.

Could you put this on a branch somewhere so I can take a look?

**Comment by jofemodo** [comment]
I will do ASAP.

FYI, we forked our version of Jalv some time ago and added the CLI to the graphic versions, so we can effectively interface the plugin with the GUI and the CLI simultaneously. We also added some extra functions to allow us to save presets from the CLI.
We have pulled your changes from time to time and it's relatively up-to-date, although not fully up-to-date. I have to merge your latest changes.

We like the CLI approach although we are considering the OSC interface because we need the fastest interaction for controlling parameters from zynthian-UI (python). Currently we use "pexpect" for CLI interaction with good results but we would like to make some tests and compare with OSC + UDS (Unix Domain Sockets). I never used UDS but i'm curious about them.

Thanks @drobilla !

**Comment by jofemodo** [comment]
I @drobilla !

I just pushed a jalv version to our zynthian fork that include an asynchronous CLI and support the property parameters. It's working smoothly with the current development version of zynthian UI.

The asynchronous CLI could be moved to OSC, but i first wanted to test the concept using simple text interface. It's working so smoothly that i'm tempted to leave it as it is. I love the simplicity and i don't see the benefit (for our use-case in zynthian) of adding more protocol layers when it works so nicely with super-simple text flowing thru stdout and stdin.

Here is the modified code:

https://github.com/zynthian/jalv/tree/asyncli

It's kind of PoC yet, but it works really fine for our use-case, including the desired property parameters and bidirectional parameter updates between native GUI and zynthian UI (before this we only had one-way updates from zynthian UI to native GUI).

All the best!

**Comment by jofemodo** [comment]
A little bit off-topic, but related ... regarding TTL presets, how can define presets using the property parameters?
I've tested something like this:

```
<https://tal-software.com/TAL-U-NO-LX-V2:preset01>
	a pset:Preset ;
	lv2:appliesTo <https://tal-software.com/TAL-U-NO-LX-V2> ;
	rdfs:label "Patch1" ;
	state:state [
	  	plug:attack "0.1"^^xsd:float ;
		plug:decay "0.4"^^xsd:float ;
		plug:sustain "0.7"^^xsd:float ;
		plug:release "0.1"^^xsd:float ;
		plug:dcolfovalue "0.7"^^xsd:float ;
		plug:volume "1.0"^^xsd:float ;
       	        plug:octavetranspose "0.0"^^xsd:float
	] .

```

(the "plug:xxx" are fully working property parameters defined in the plugin's TTL)
Is this OK? Perhaps it's LV2-compliant but not supported/implemented in jalv?
I've tested a lot of ways and variations but i don't get it to work.

BTW, do you have a zynthian device? I can't remember ...

Thanks a lot!

**Comment by jofemodo** [comment]
Happy New Year, @drobilla!

**Comment by drobilla** [comment]
Thanks.  I'll take a look at that shortly.  I have been doing a serious overhaul of the internals to better support this and similar things, but I've been on a break.  I'll probably get back to it next week. 

Unfortunately this will definitely have massive conflicts with any changes you've made to the existing code, but there's now some clean separation between the audio thread, main app, and (to a lesser extent) UIs with a clean way to communicate between them which I think is a necessary prerequisite for this (and other features way beyond what the original monolithic toy that jalv started life as was designed for).  The parameters part isn't quite done yet though.

**Comment by jofemodo** [comment]
Woaoo! I'm impatient ... although as you say, it's going to be a hard task to merge your changes.

BTW, do you have a zynthian device? If you are interested i could rescue a prototype unit from our lab and ship it to you.

Regards,

**Comment by jofemodo** [comment]
Please, could you give me a hint about the "property parameter presets" question?
I'm really blocked on this.

**Comment by drobilla** [comment]
> I'm impatient
> I'm really blocked on this

A few comments above, I asked you for a link to code that supposedly already existed.  You didn't respond at all for over two weeks, and then without the code anyway.  Then, a whole other month later, you finally linked me to something.  Immediately after doing so, you started abusing this ticket and nagging me - right in the middle of the holiday season no less - because apparently _I_ am expected to have a turnaround time of a couple days?  When I'm not even home, let alone working?

Yeah, no.  Consider this your first and final warning, I don't take kindly to people who act like they're entitled to my time.  If you want to treat me like your employee, you're welcome to _pay me_.  Otherwise, I will get around to things if and when I feel like it.

> BTW, do you have a zynthian device? I can't remember

No, I didn't even know it existed until now.

> Please, could you give me a hint about the "property parameter presets" question?

Plugins are responsible for saving and restoring their own state.  There is nothing special about parameters here.

**Comment by jofemodo** [comment]
Sorry @drobilla !

I didn't intend to annoy you in any way. Of course,i don't pretend "you have to help me" or "i have the right to your attention". If some of my comments suggest any kind of exigence, please, excuse my words as it's not my intention at all.

I hope you have enjoyed your holidays with family and friends, same i enjoyed mine. Really! I simply answered when i had the time/focus and don't expect nothing different from you.

I suppose you could imagine, but perhaps it's good you read it. My acknowledgment and admiration for your work is HUGE. My work and vital project, zynthian, ride on giant shoulders, being you one of these giants. LV2 is a keystone of zynthian project and as zynthian is reaching more and more people, also does your work.

BTW, i'm happy to know you finally discovered our little project zynthian. It's good to know we have just scrapped the surface ;-)

I hope my words about "sending a zynthian device" to you didn't sound to your ears like i'm "treating you as an employee". More in the contrary, it's just my way of showing some gratitude and perhaps, also a little blinking to my own vanity with the little hope you appreciate zynthian.

At a first glance, perhaps you could think zynthian is something similar to MOD-devices, etc. but i would like you perceive the subtle philosophical differences and the true beauty of our little project. Probably the best way is you take a look to this thread:

https://discourse.zynthian.org/t/success-cases/41

You can keep scrolling for a while! It's a very long thread and it covers 8 years of DIY zynthian builds. Most zynthians shown in this thread are built from scratch by brave tinkerers who love to build their own music gear.

OK! Excuse me again if i wrote too much ...

All the best,

**Comment by drobilla** [comment]
It's fine.  I've been in this game for a while, I know that entitled behaviour needs to be nipped in the bud.  I'm actively hostile to anything approaching "nagging" on purpose to dis-incentivize it.  It has nothing to do with the device or project: in this society of entitled consumers, FLOSS developers (among many others) get treated like corporations, as if we somehow owe people even more free labour.  I just happen to be at a point in my life where I have absolutely zero tolerance for that nonsense anymore.

As for hardware, I appreciate the offer, but at the moment I really don't have the capacity to swap in a whole new project and hardware device to interact with, so it would probably just be wasted at the moment.  As my near-infinite TODO list of chores gets less overwhelming, maybe I'll change my mind and let you know.

At the moment, I am working on releasing new versions of all the LV2 libraries, a Jalv release with the control improvements among many other things will follow shortly (ideally by next week, almost certainly within the month).

**Comment by drobilla** [comment]
Implemented (mostly) in 5bc5b73.

The command-line interface itself is still pretty toyish (if less so), but the program now has a frontend-independent control system that supports all common/standard types of controls, and they can be investigated and manipulated in the console interface.
